### ISSUE TITLE
Unable to locate C library

### ISSUE THREAD

**Issue by letorbi** [chrisjbillington/inotify_simple#21]
Hej Chris,

one of the users of Synoindex Watcher reported the following error in https://github.com/letorbi/synoindexwatcher/issues/38:

`root@syno:~# python -m synoindexwatcher 2020-07-30 12:51:08,326 ERROR An uncaught exception occurred Traceback (most recent call last): File "/usr/lib/python2.7/site-packages/synoindexwatcher/__main__.py", line 224, in <module> start() File "/usr/lib/python2.7/site-packages/synoindexwatcher/__main__.py", line 189, in start inotify = INotify() File "/usr/lib/python2.7/site-packages/inotifyrecursive/inotifyrecursive.py", line 31, in __init__ inotify_simple.INotify.__init__(self) File "/usr/lib/python2.7/site-packages/inotify_simple.py", line 84, in __init__ global _libc; _libc = _libc or CDLL(find_library('c'), use_errno=True) File "/usr/lib/python2.7/ctypes/util.py", line 242, in find_library raise RuntimeError("can not find library %s" % name) RuntimeError: can not find library c`

After some research I've found out that the error only occurs when inotify-simple 1.3.4 is installed. With all previous versions everything works without any problems. Since you've recently changed the way to load *libc.so* (#20), I assume the problem is related to this change.

The affected systems seem to be Synology DiskStations, here's the `uname -a` output of mine:

```
Linux xxxxxx 3.2.40 #25426 Wed Jul 8 03:15:48 CST 2020 armv7l GNU/Linux synology_armada370_213j
```

The output of `ls -la /usr/lib/libc.*` is:

```
lrwxrwxrwx 1 root root 20 Jul 31 14:08 /usr/lib/libc.so.6 -> libc-2.20-2014.11.so
```

I was unable to locate *inotify.h* on the system.

Please write me, if you need more info to fix the problem.

Bye,
Torben

**Comment by letorbi** [comment]
For what it's worth, I've just tried to read the values of `oct(os.O_CLOEXEC)` and `oct(os.O_NONBLOCK)` on my DiskStation:

```
Python 2.7.12 (default, May  7 2020, 04:26:53) 
[GCC 4.9.3 20150311 (prerelease)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> oct(os.O_CLOEXEC)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'O_CLOEXEC'
>>> oct(os.O_NONBLOCK)
'04000'
```

It seems that `os.O_CLOEXEC` does not even exist at all. which is no surprise, since [it has been added with Python 3.3](https://docs.python.org/3.8/library/os.html#os.O_CLOEXEC). So this might be the root of this issue.

However, Synology users are stuck with Python 2.7.12, if they don't want to install any third party packages. Is there a chance that you'll try to find a Python 2.7.12 compatible solution to load *libc.so* or would you rather just drop support for Python 2.7?

**Comment by chrisjbillington** [comment]
`O_CLOEXEC` not being defined is fine, it seems its not able to find libc.

That's annoying, as I changed it only because someone else claimed their system was *not* able to find libc under the name `libc.so.6`.

When I google "RuntimeError: can not find library c" I only get results about Synology NASs, lol.

In fact, my copy of `/usr/lib/python2.7/ctypes/util.py` (Python 2.7.18) does not contain any lines like `raise RuntimeError("can not find library %s" % name)`.

`find_library` is supposed to return `None` if it cannot find the library, not raise an exception. I'm very confused.

I can work around by hard-coding `libc.so.6` if find_library fails, but I'd like to know why it's failing with a RuntimeError instead of returning None. Might git blame `ctypes/util.py` to find out if anything changed between python 2.7.12 and 2.7.18

**Comment by letorbi** [comment]
Thanks for the fast reply. The 2.7.12 version of *ctypes/util.py* throws a `RuntimeException` indeed:

```
def find_library(name):
    if not _findSoname_ldconfig(name) and not _get_soname(_findLib_gcc(name)):
        raise RuntimeError("can not find library %s" % name)
```

> When I google "RuntimeError: can not find library c" I only get results about Synology NASs, lol.

Yeah, Synology's Linux distribution has so many quirks, I'm not even sure, if you could call it a Linux anymore ;) I've actually installed an Alpine Linux in a chroot-environment to have a decent system...

**Comment by chrisjbillington** [comment]
CPython 2.7.12 does not contain that RuntimeError (and neither does any other version, as far as I can tell):

https://github.com/python/cpython/blob/v2.7.12/Lib/ctypes/util.py#L236

It must be a patched CPython, which would explain why googling the error only turns up Synology-related results.

I'll make it work anyway, presumably this change will fix it?

```diff
diff --git a/inotify_simple.py b/inotify_simple.py
index 6558288..2d7664e 100644
--- a/inotify_simple.py
+++ b/inotify_simple.py
@@ -81,7 +81,11 @@ class INotify(FileIO):
                 reads of the file descriptor (for example if the application reads data
                 manually with ``os.read(fd)``) to raise ``BlockingIOError`` if no data
                 is available."""
-        global _libc; _libc = _libc or CDLL(find_library('c'), use_errno=True)
+        try:
+            libc_so = find_library('c')
+        except RuntimeError: # Python on Synology NASs raises a RuntimeError
+            libc_so = None
+        global _libc; _libc = _libc or CDLL(libc_so or 'libc.so.6', use_errno=True)
         O_CLOEXEC = getattr(os, 'O_CLOEXEC', 0) # Only defined in Python 3.3+
         flags = (not inheritable) * O_CLOEXEC | bool(nonblocking) * os.O_NONBLOCK
         FileIO.__init__(self, _libc_call(_libc.inotify_init1, flags), mode='rb')
```

**Comment by letorbi** [comment]
I've just tested the patch: Everything works fine after is has been applied.

BTW: Had to apply it manually because the Synology system does not offer a `patch` command :astonished:

**Comment by chrisjbillington** [comment]
Great! Do you know if the same problem exists in Python 3? If not it would be nice to know so I can cull compatibility code like this when I eventually drop python2 support.

**Comment by letorbi** [comment]
Synology DiskStations do not have Python 3 (yet another weirdness). There are some third-party packages that provide it, but they have to be installed manually. I assume that these packages simply provide a vanilla Python, but I cannot say for sure.

**Comment by chrisjbillington** [comment]
Cool, well if they ever ship Python 3 we can think about it then.

I've released 1.3.5 with the fix
